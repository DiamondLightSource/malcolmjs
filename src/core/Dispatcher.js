/*
 * React.js Starter Kit
 * Copyright (c) 2014 Konstantin Tarkus (@koistya), KriaSoft LLC.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.txt file in the root directory of this source tree.
 */

import Flux from 'flux';
import PayloadSources from '../constants/PayloadSources';
import assign from 'react/lib/Object.assign';

/**
 * A singleton that operates as the central hub for application updates.
 * For more information visit https://facebook.github.io/flux/
 */
let Dispatcher = assign(new Flux.Dispatcher(), {

  /**
   * @param {object} action The details of the action, including the action's
   * type and additional data coming from the server.
   */
  handleServerAction(action) {
    var payload = {
      source: PayloadSources.SERVER_ACTION,
      action: action
    };
    this.dispatch(payload);
  },

  /**
   * @param {object} action The details of the action, including the action's
   * type and additional data coming from the view.
   */
  handleViewAction(action) {
    var payload = {
      source: PayloadSources.VIEW_ACTION,
      action: action
    };
    this.dispatch(payload);
  }

});

export default Dispatcher;
