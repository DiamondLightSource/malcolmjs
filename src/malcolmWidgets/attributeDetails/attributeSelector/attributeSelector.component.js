import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from 'material-ui/styles';
import { darken } from 'material-ui/styles/colorManipulator';
import { connect } from 'react-redux';
import WidgetLED from '../../led/widgetLED.component';
import WidgetCheckbox from '../../checkbox/checkbox.component';
import TextUpdate from '../../textUpdate/WidgetTextUpdate.component';
import {
  malcolmPutAction,
  malcolmSetPending,
} from '../../../malcolm/malcolmActionCreators';

const styles = theme => ({
  spinner: {
    size: 44,
    color: darken(theme.palette.primary.light, 0.25),
  },
});

const AttributeSelector = props => {
  if (props.attribute && props.attribute.meta && props.attribute.meta.tags) {
    const { tags } = props.attribute.meta;
    const widgetTagIndex = tags.findIndex(t => t.indexOf('widget:') !== -1);

    if (widgetTagIndex !== -1) {
      if (tags[widgetTagIndex] === 'widget:led') {
        return (
          <WidgetLED
            LEDState={props.attribute.value}
            colorBorder={props.theme.palette.primary.light}
            colorCenter={props.theme.palette.primary.light}
          />
        );
      } else if (tags[widgetTagIndex] === 'widget:textupdate') {
        return <TextUpdate Text={props.attribute.value} />;
      }
    }

    if (tags.some(t => t === 'widget:checkbox')) {
      return (
        <WidgetCheckbox
          CheckState={props.attribute.value}
          Pending={props.attribute.pending}
          checkEventHandler={isChecked =>
            props.checkHandler(props.attribute.path, isChecked)
          }
        />
      );
    }
  }
  return <div>Hello</div>;
};

const mapStateToProps = () => ({});

const mapDispatchToProps = dispatch => ({
  checkHandler: (path, isChecked) => {
    dispatch(malcolmSetPending(path));
    dispatch(malcolmPutAction(path, isChecked));
  },
});

AttributeSelector.propTypes = {
  attribute: PropTypes.shape({
    meta: PropTypes.shape({
      tags: PropTypes.arrayOf(PropTypes.string),
    }),
    path: PropTypes.string,
    value: PropTypes.oneOfType([
      PropTypes.bool,
      PropTypes.number,
      PropTypes.string,
    ]),
    pending: PropTypes.bool,
  }).isRequired,
  theme: PropTypes.shape({
    palette: PropTypes.shape({
      primary: PropTypes.shape({
        light: PropTypes.string,
      }),
      secondary: PropTypes.shape({
        main: PropTypes.string,
      }),
    }),
  }).isRequired,
  checkHandler: PropTypes.func.isRequired,
};

export default connect(mapStateToProps, mapDispatchToProps)(
  withStyles(styles, { withTheme: true })(AttributeSelector)
);
