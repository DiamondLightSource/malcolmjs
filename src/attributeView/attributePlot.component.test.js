import React from 'react';
import { createShallow } from '@material-ui/core/test-utils';
import MockCircularBuffer from '../malcolm/reducer/attribute.reducer.mocks';
import { malcolmTypes } from '../malcolmWidgets/attributeDetails/attributeSelector/attributeSelector.component';
import AttributePlot from './attributePlot.component';

describe('attributePlot', () => {
  let shallow;
  let mockArchive;

  beforeEach(() => {
    shallow = createShallow({ dive: true });
    mockArchive = {
      timeStamp: new MockCircularBuffer(5),
      plotValue: new MockCircularBuffer(5),
      meta: {},
    };
  });

  it('renders correctly for number', () => {
    mockArchive.meta.typeid = malcolmTypes.number;
    mockArchive.timeStamp.push(new Date(1000));
    mockArchive.timeStamp.push(new Date(2000));
    mockArchive.timeStamp.push(new Date(3000));
    mockArchive.plotValue.push(1);
    mockArchive.plotValue.push(-1);
    mockArchive.plotValue.push(2);

    const wrapper = shallow(<AttributePlot attribute={mockArchive} />);
    expect(wrapper.dive()).toMatchSnapshot();
  });

  it('renders correctly for bool', () => {
    mockArchive.meta.typeid = malcolmTypes.bool;
    mockArchive.timeStamp.push(new Date(1000));
    mockArchive.timeStamp.push(new Date(2000));
    mockArchive.timeStamp.push(new Date(3000));
    mockArchive.plotValue.push(1);
    mockArchive.plotValue.push(0);
    mockArchive.plotValue.push(1);

    const wrapper = shallow(<AttributePlot attribute={mockArchive} />);
    expect(wrapper.dive()).toMatchSnapshot();
  });
});
