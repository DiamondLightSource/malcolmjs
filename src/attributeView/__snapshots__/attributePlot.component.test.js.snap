// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`attributePlot renders correctly for bool 1`] = `
<div
  style={
    Object {
      "height": "100%",
      "width": "100%",
    }
  }
/>
`;

exports[`attributePlot renders correctly for number 1`] = `
<div
  style={
    Object {
      "height": "100%",
      "width": "100%",
    }
  }
/>
`;
